group_theory/perm.lean

```diff
@@ -286,9 +385,204 @@ funext $ λ f,
 let ⟨l, hl₁, hl₂⟩ := trunc.out (trunc_swap_factors f) in
 have hsl : ∀ a ∈ l.map s, a = (-1 : units ℤ) := λ a ha,
   let ⟨g, hg⟩ := list.mem_map.1 ha in hg.2 ▸  this (hl₂ _ hg.1),
-have hsignl :  ∀ a ∈ l.map sign, a = (-1 : units ℤ) := λ a ha,
-  let ⟨g, hg⟩ := list.mem_map.1 ha in hg.2 ▸ sign_eq_of_is_swap (hl₂ _ hg.1),
-by rw [← hl₁, is_group_hom.prod s, is_group_hom.prod (@sign α _ _),
-  list.eq_repeat'.2 hsl, list.eq_repeat'.2 hsignl, list.length_map, list.length_map]
+by rw [← hl₁, is_group_hom.prod s, list.eq_repeat'.2 hsl, list.length_map,
+     list.prod_repeat, hl₁, sign_eq_pow_length_swap_factors hl₁ hl₂]
+
+lemma sign_subtype_perm [fintype α] (f : perm α) {p : α → Prop} [decidable_pred p]
+  (h₁ : ∀ x, p x ↔ p (f x)) (h₂ : ∀ x, f x ≠ x → p x) : sign (subtype_perm f h₁) = sign f :=
+let l := trunc.out (trunc_swap_factors (subtype_perm f h₁)) in
+have hl' : ∀ g' ∈ l.1.map of_subtype, is_swap g' :=
+  λ g' hg',
+  let ⟨g, hg⟩ := list.mem_map.1 hg' in
+  hg.2 ▸ is_swap_of_subtype (l.2.2 _ hg.1),
+have hl'₂ : (l.1.map of_subtype).prod = f,
+  by rw [← is_group_hom.prod of_subtype l.1, l.2.1, of_subtype_subtype_perm _ h₂],
+by rw [sign_eq_pow_length_swap_factors l.2.1 l.2.2, sign_eq_pow_length_swap_factors hl'₂ hl'];
+  simp
+
+@[simp] lemma sign_of_subtype [fintype α] {p : α → Prop} [decidable_pred p]
+  (f : perm (subtype p)) : sign (of_subtype f) = sign f :=
+have ∀ x, of_subtype f x ≠ x → p x, from λ x, not_imp_comm.1 (of_subtype_apply_of_not_mem f),
+by conv {to_rhs, rw [← subtype_perm_of_subtype f, sign_subtype_perm _ _ this]}
+
+lemma sign_eq_sign_of_equiv [fintype α] [decidable_eq β] [fintype β] (f : perm α) (g : perm β)
+  (e : α ≃ β) (h : ∀ x, e (f x) = g (e x)) : sign f = sign g :=
+have hg : g = (e.symm.trans f).trans e, from equiv.ext _ _ $ by simp [h],
+by rw [hg, sign_symm_trans_trans]
+
+lemma sign_bij [fintype α] [decidable_eq β] [fintype β]
+  (f : perm α) (g : perm β) (i : Π x : α, f x ≠ x → β)
+  (h : ∀ x hx hx', i (f x) hx' = g (i x hx))
+  (hi : ∀ x₁ x₂ hx₁ hx₂, i x₁ hx₁ = i x₂ hx₂ → x₁ = x₂)
+  (hg : ∀ y, g y ≠ y → ∃ x hx, i x hx = y) :
+  sign f = sign g :=
+calc sign f = sign (@subtype_perm _ f (λ x, f x ≠ x) (by simp)) :
+eq.symm (sign_subtype_perm _ _ (λ _, id))
+... = sign (@subtype_perm _ g (λ x, g x ≠ x) (by simp)) :
+sign_eq_sign_of_equiv _ _
+  (equiv.of_bijective
+    (show function.bijective (λ x : {x // f x ≠ x},
+      (⟨i x.1 x.2, have f (f x) ≠ f x, from mt (λ h, f.bijective.1 h) x.2,
+        by rw [← h _ x.2 this]; exact mt (hi _ _ this x.2) x.2⟩ : {y // g y ≠ y})),
+      from ⟨λ ⟨x, hx⟩ ⟨y, hy⟩ h, subtype.eq (hi _ _ _ _ (subtype.mk.inj h)),
+        λ ⟨y, hy⟩, let ⟨x, hfx, hx⟩ := hg y hy in ⟨⟨x, hfx⟩, subtype.eq hx⟩⟩))
+    (λ ⟨x, _⟩, subtype.eq (h x _ _))
+... = sign g : sign_subtype_perm _ _ (λ _, id)
+
+lemma is_cycle_swap {x y : α} (hxy : x ≠ y) : is_cycle (swap x y) :=
+⟨y, by rwa swap_apply_right,
+λ a (ha : ite (a = x) y (ite (a = y) x a) ≠ a),
```

Comment:
indent the lines inside the anonymous constructor. (but not the `else`. I don't think we have a general rule, but to be consistent with your indentation style, it should be at the same column as `if`.)